<font size="7"><p style="text-align: center;">**Gérer ses données avec python**</font></p>
<br><br><br>
<img src="img\pandas.svg">

## <u>1. Importer des données avec Pandas</u>
Pandas offre une panoplie de connecteurs pour importer des données (CSV, Excel, SQL, hdfs, etc.).

data source: https://www.kaggle.com/rashikrahmanpritom/heart-attack-analysis-prediction-dataset

In [25]:
!pip install pyodbc

Defaulting to user installation because normal site-packages is not writeable


In [26]:
!pip install lxml

Defaulting to user installation because normal site-packages is not writeable


In [27]:
!pip install pandas

Defaulting to user installation because normal site-packages is not writeable


In [28]:
!uv add sqlalchemy

Resolved 47 packages in 1ms
Checked 41 packages in 4ms


In [29]:
%pip install requests

Note: you may need to restart the kernel to use updated packages.


c:\Users\e.vervaeren\Desktop\Learning-IOT-AI-Architect\Python\pyhton_data_science\.venv\Scripts\python.exe: No module named pip


In [30]:
import numpy as np
import pandas as pd

import requests as rq
from io import StringIO
from sqlalchemy import create_engine

In [31]:
df = pd.read_csv("heart.csv")

# La méthode head() permet de n'afficher que les premières lignes (5 par défaut)
# Ici on en affiche 10
display(df.head(10))
display(df.tail())

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1
5,57,1,0,140,192,0,1,148,0,0.4,1,0,1,1
6,56,0,1,140,294,0,0,153,0,1.3,1,0,2,1
7,44,1,1,120,263,0,1,173,0,0.0,2,0,3,1
8,52,1,2,172,199,1,1,162,0,0.5,2,0,3,1
9,57,1,2,150,168,0,1,174,0,1.6,2,0,2,1


,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output
298,57,0,0,140,241,0,1,123,1,0.2,1,0,3,0
299,45,1,3,110,264,0,1,132,0,1.2,1,0,3,0
300,68,1,0,144,193,1,1,141,0,3.4,1,2,3,0
301,57,1,0,130,131,0,1,115,1,1.2,1,1,3,0
302,57,0,1,130,236,0,0,174,0,0.0,1,1,2,0


Il existe 2 structures de donnée dans pandas:
- les **Series**: correspond à une colonne, à un numpy array avec un type de donnée unique
- les **DataFrame**: correspond à une collection de **Series**

In [32]:
display(type(df))
type(df['age'])

pandas.DataFrame

pandas.Series

Un DataFrame possède plusieurs attributs que l'on peut utiliser au besoin

In [33]:
print(df.shape)
print(50*'-')
print(df.size)
print(50*'-')
print(df.dtypes)
print(50*'-')
print(df.columns)
print(50*'-')
print(df.index)
print(50*'-')
print(df.values)

(303, 14)
--------------------------------------------------
4242
--------------------------------------------------
age           int64
sex           int64
cp            int64
trtbps        int64
chol          int64
fbs           int64
restecg       int64
thalachh      int64
exng          int64
oldpeak     float64
slp           int64
caa           int64
thall         int64
output        int64
dtype: object
--------------------------------------------------
Index(['age', 'sex', 'cp', 'trtbps', 'chol', 'fbs', 'restecg', 'thalachh',
       'exng', 'oldpeak', 'slp', 'caa', 'thall', 'output'],
      dtype='str')
--------------------------------------------------
RangeIndex(start=0, stop=303, step=1)
--------------------------------------------------
[[63.  1.  3. ...  0.  1.  1.]
 [37.  1.  2. ...  0.  2.  1.]
 [41.  0.  1. ...  0.  2.  1.]
 ...
 [68.  1.  0. ...  2.  3.  0.]
 [57.  1.  0. ...  1.  3.  0.]
 [57.  0.  1. ...  1.  2.  0.]]


Méthode info() : La méthode info() nous renvoi un condensé d'informations utiles concernant notre dataframe

In [34]:
df.info() # Must know

<class 'pandas.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trtbps    303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalachh  303 non-null    int64  
 8   exng      303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slp       303 non-null    int64  
 11  caa       303 non-null    int64  
 12  thall     303 non-null    int64  
 13  output    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


## <u>2. Comment manipuler ses données avec pandas ?</u>

**2.1. Indexation et colonnes**

In [35]:
df[['age','cp']].head()


,age,cp
0,63,3
1,37,2
2,41,1
3,56,1
4,57,0


In [36]:
df['age'].head()

df["age"].head() # => Ceci est un dataframe car " "

0    63
1    37
2    41
3    56
4    57
Name: age, dtype: int64

In [37]:
df.age

0      63
1      37
2      41
3      56
4      57
       ..
298    57
299    45
300    68
301    57
302    57
Name: age, Length: 303, dtype: int64

In [38]:
test = pd.DataFrame({'prenom':["Georges","Jules","Max"],
                     'age':[20,25,30]},index=['a','b','c'])
display(test)
test.loc['a':'c',['prenom','age']]

,prenom,age
a,Georges,20
b,Jules,25
c,Max,30


,prenom,age
a,Georges,20
b,Jules,25
c,Max,30


In [39]:
df.loc[10:20,["age","cp"]]   # dF.loc[indexDebut:indexFin, nomattribut]  loc est inclusif

,age,cp
10,54,0
11,48,2
12,49,1
13,64,3
14,58,3
15,50,2
16,58,2
17,66,3
18,43,0
19,69,3


In [40]:
df.iloc[5:10,0:5] # index ligne , index colonne

,age,sex,cp,trtbps,chol
5,57,1,0,140,192
6,56,0,1,140,294
7,44,1,1,120,263
8,52,1,2,172,199
9,57,1,2,150,168


In [41]:
df.at[0,'age']

np.int64(63)

In [42]:
df.iat[0,0]

np.int64(63)

In [43]:
# Avec le slicing, comme pour les listes python
# En ne sélectionnant qu'une seule ligne
data_sliced = df.loc[[10]]
display(data_sliced)
# Selectionner plusieurs lignes avec une list ou une array
data_sliced2 = df.loc[[10,20,30]]
display(data_sliced2)
# Selectionner sur base de l'index et des noms de colonnes
data_sliced3 = df.loc[[10,20,30],['age','chol']]
display(data_sliced3)
data_sliced4 = df[['age','chol']].loc[1:5]
display(data_sliced4)
print(df.age)

display(data_sliced3.dtypes)
data_sliced4.dtypes

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output
10,54,1,0,140,239,0,1,160,0,1.2,2,0,2,1


,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output
10,54,1,0,140,239,0,1,160,0,1.2,2,0,2,1
20,59,1,0,135,234,0,1,161,0,0.5,1,0,3,1
30,41,0,1,105,198,0,1,168,0,0.0,2,1,2,1


,age,chol
10,54,239
20,59,234
30,41,198


,age,chol
1,37,250
2,41,204
3,56,236
4,57,354
5,57,192


0      63
1      37
2      41
3      56
4      57
       ..
298    57
299    45
300    68
301    57
302    57
Name: age, Length: 303, dtype: int64


age     int64
chol    int64
dtype: object

age     int64
chol    int64
dtype: object

In [44]:
df[1:10]

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1
5,57,1,0,140,192,0,1,148,0,0.4,1,0,1,1
6,56,0,1,140,294,0,0,153,0,1.3,1,0,2,1
7,44,1,1,120,263,0,1,173,0,0.0,2,0,3,1
8,52,1,2,172,199,1,1,162,0,0.5,2,0,3,1
9,57,1,2,150,168,0,1,174,0,1.6,2,0,2,1


In [45]:
df.loc[:0,'age':'cp']

,age,sex,cp
0,63,1,3


In [46]:
# En sélectionnant un range
data_sliced = df.loc[74:80]
data_sliced

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output
74,43,0,2,122,213,0,1,165,0,0.2,1,0,2,1
75,55,0,1,135,250,0,0,161,0,1.4,1,0,2,1
76,51,1,2,125,245,1,0,166,0,2.4,1,0,2,1
77,59,1,1,140,221,0,1,164,1,0.0,2,0,2,1
78,52,1,1,128,205,1,1,184,0,0.0,2,0,2,1
79,58,1,2,105,240,0,0,154,1,0.6,1,0,3,1
80,41,1,2,112,250,0,1,179,0,0.0,2,0,2,1


In [47]:
df.iloc[5:8,[1,5]]

,sex,fbs
5,1,0
6,0,0
7,1,0


In [48]:
df.at[0,'age']

np.int64(63)

In [49]:
df.iat[0,0]

np.int64(63)

In [50]:
# Il est également possible de ne sélectionner qu'une colonne en particulier
# Avec les crochets et la colonne voulue
data_sliced = df[['age','chol']]
data_sliced
data_chol = df['chol']
print (data_sliced)
data_chol

     age  chol
0     63   233
1     37   250
2     41   204
3     56   236
4     57   354
..   ...   ...
298   57   241
299   45   264
300   68   193
301   57   131
302   57   236

[303 rows x 2 columns]


0      233
1      250
2      204
3      236
4      354
      ... 
298    241
299    264
300    193
301    131
302    236
Name: chol, Length: 303, dtype: int64

In [51]:
# Avec plusieurs colonnes
# A noter les doubles crochets
data_sliced = df[['age','output']].loc[54:60]
data_sliced

,age,output
54,63,1
55,52,1
56,48,1
57,45,1
58,34,1
59,57,1
60,71,1


**2.2. Boolean indexing**

Nous avons vu jusqu'ici comment sélectionner des parties de notre DataFrame avec les indexes ou le nom des colonnes.

Nous pourrions également faire une sélection basée sur des conditions. 

C'est ce que l'on appelle le **boolean indexing**.

In [52]:
df['age']>58

0       True
1      False
2      False
3      False
4      False
       ...  
298    False
299    False
300     True
301    False
302    False
Name: age, Length: 303, dtype: bool

In [53]:
# On ne sélectionne que les personnes plus agées que 58 ans
data_sliced = df[df['age'] > 58]
data_sliced
data_sliced2= df[df['age']>58].loc[0:50]
display(data_sliced2)
print(data_sliced2.loc[38])
print(data_sliced2.iloc[10])

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1
13,64,1,3,110,211,0,0,144,1,1.8,1,0,2,1
17,66,0,3,150,226,0,1,114,0,2.6,0,0,2,1
19,69,0,3,140,239,0,1,151,0,1.8,2,2,2,1
20,59,1,0,135,234,0,1,161,0,0.5,1,0,3,1
23,61,1,2,150,243,1,1,137,1,1.0,1,0,2,1
25,71,0,1,160,302,0,1,162,0,0.4,2,2,2,1
26,59,1,2,150,212,1,1,157,0,1.6,2,0,2,1
28,65,0,2,140,417,1,0,157,0,0.8,2,1,2,1
31,65,1,0,120,177,0,1,140,0,0.4,2,0,3,1


age          65.0
sex           0.0
cp            2.0
trtbps      155.0
chol        269.0
fbs           0.0
restecg       1.0
thalachh    148.0
exng          0.0
oldpeak       0.8
slp           2.0
caa           0.0
thall         2.0
output        1.0
Name: 38, dtype: float64
age          65.0
sex           0.0
cp            2.0
trtbps      155.0
chol        269.0
fbs           0.0
restecg       1.0
thalachh    148.0
exng          0.0
oldpeak       0.8
slp           2.0
caa           0.0
thall         2.0
output        1.0
Name: 38, dtype: float64


On peut également mettre plusieurs conditions.

**Attention:** les opérateurs logiques "and" et "or" classiques de Python ne fonctionnent pas.

Il faut impérativement utiliser les opérateurs **&** et **|** et mettre des **()** autour des conditions.

In [54]:
data_sliced2 = df[(df['age'] > 58) & (df['output'] == 1)]
data_sliced2.head()

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1
13,64,1,3,110,211,0,0,144,1,1.8,1,0,2,1
17,66,0,3,150,226,0,1,114,0,2.6,0,0,2,1
19,69,0,3,140,239,0,1,151,0,1.8,2,2,2,1
20,59,1,0,135,234,0,1,161,0,0.5,1,0,3,1


Il est possible de créer de nouvelles colonnes de manière à enrichir vos données.

In [55]:
# Ici on souhaite créer une colonne basée sur la colonne age de manière à obtenir des catégories
# On l'instancie d'abord à 'unknown'
df['age_cat'] = np.nan
display(df.head())
del df['age_cat']

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output,age_cat
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1,NaN
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1,NaN
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1,NaN
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1,NaN
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1,NaN


In [56]:
df['newcol'] = df['age']*df['oldpeak']
display(df.head())
df.drop(columns=['newcol'],inplace=True) #inplace True permetr d'altérer
#df = df.drop(columns=['newcol'])
#df.head()
df

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output,newcol
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1,144.9
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1,129.5
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1,57.4
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1,44.8
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1,34.2


,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
298,57,0,0,140,241,0,1,123,1,0.2,1,0,3,0
299,45,1,3,110,264,0,1,132,0,1.2,1,0,3,0
300,68,1,0,144,193,1,1,141,0,3.4,1,2,3,0
301,57,1,0,130,131,0,1,115,1,1.2,1,1,3,0


In [57]:
# Nous allons utiliser le boolean indexing pour modifier la catégorie
# Attention à utiliser .loc lorsque l'on chercher un modifier des valeurs
df.loc[df['age'] <= 18, 'age_cat'] = 'child'
display(df)
df.loc[(df['age'] > 18) & (df['age'] < 66), 'age_cat'] = 'adult'
df.loc[df['age'] > 65, 'age_cat'] = 'senior'
df

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output,age_cat
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1,NaN
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1,NaN
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1,NaN
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1,NaN
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
298,57,0,0,140,241,0,1,123,1,0.2,1,0,3,0,NaN
299,45,1,3,110,264,0,1,132,0,1.2,1,0,3,0,NaN
300,68,1,0,144,193,1,1,141,0,3.4,1,2,3,0,NaN
301,57,1,0,130,131,0,1,115,1,1.2,1,1,3,0,NaN


,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output,age_cat
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1,adult
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1,adult
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1,adult
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1,adult
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1,adult
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
298,57,0,0,140,241,0,1,123,1,0.2,1,0,3,0,adult
299,45,1,3,110,264,0,1,132,0,1.2,1,0,3,0,adult
300,68,1,0,144,193,1,1,141,0,3.4,1,2,3,0,senior
301,57,1,0,130,131,0,1,115,1,1.2,1,1,3,0,adult


## <u>3. Opérations mathématiques</u>
- Pandas supporte bien évidemment toutes les opérations mathématiques de base
- Pandas supporte également toutes les opérations de numpy

In [58]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 15 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trtbps    303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalachh  303 non-null    int64  
 8   exng      303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slp       303 non-null    int64  
 11  caa       303 non-null    int64  
 12  thall     303 non-null    int64  
 13  output    303 non-null    int64  
 14  age_cat   303 non-null    str    
dtypes: float64(1), int64(13), str(1)
memory usage: 35.6 KB


In [59]:
data_math = df.copy()
# Addition d'une constante à l'entièreté d'une colonne
data_math['age'] += 100
# identique a data_math['age'] = data_math['age'] +100
data_math

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output,age_cat
0,163,1,3,145,233,1,0,150,0,2.3,0,0,1,1,adult
1,137,1,2,130,250,0,1,187,0,3.5,0,0,2,1,adult
2,141,0,1,130,204,0,0,172,0,1.4,2,0,2,1,adult
3,156,1,1,120,236,0,1,178,0,0.8,2,0,2,1,adult
4,157,0,0,120,354,0,1,163,1,0.6,2,0,2,1,adult
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
298,157,0,0,140,241,0,1,123,1,0.2,1,0,3,0,adult
299,145,1,3,110,264,0,1,132,0,1.2,1,0,3,0,adult
300,168,1,0,144,193,1,1,141,0,3.4,1,2,3,0,senior
301,157,1,0,130,131,0,1,115,1,1.2,1,1,3,0,adult


In [60]:
# On peut également additionner les colonnes entre elles
data_math['age'] = data_math['age']+data_math['chol']

On peut retrouver toutes les fonctions mathématiques de numpy ici:

https://numpy.org/doc/stable/reference/routines.math.html

In [61]:
data_math['age'] = np.log1p(data_math['age'])
data_math
data_test = np.sum(data_math['age'])
data_test

np.float64(1814.005681170679)

## <u>4. Agrégations et statistiques</u>
Tout comme avec numpy, il est possible d'agréger nos données selon un axe spécifique.

Pandas fournit également quelque statistiques de base.

In [62]:
# Par défaut axis=0
df.sum(axis=0)

age                                                     16473
sex                                                       207
cp                                                        293
trtbps                                                  39882
chol                                                    74618
fbs                                                        45
restecg                                                   160
thalachh                                                45343
exng                                                       99
oldpeak                                                 315.0
slp                                                       424
caa                                                       221
thall                                                     701
output                                                    165
age_cat     adultadultadultadultadultadultadultadultadulta...
dtype: object

In [63]:
df.sum(axis=1,numeric_only=True)  # numeric_only=True

0      600.3
1      614.5
2      554.4
3      598.8
4      701.6
       ...  
298    567.2
299    561.2
300    558.4
301    442.2
302    602.0
Length: 303, dtype: float64

In [64]:
df['age'].astype(float)

0      63.0
1      37.0
2      41.0
3      56.0
4      57.0
       ... 
298    57.0
299    45.0
300    68.0
301    57.0
302    57.0
Name: age, Length: 303, dtype: float64

In [65]:
df.head()

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output,age_cat
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1,adult
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1,adult
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1,adult
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1,adult
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1,adult


In [66]:
# Imaginons que nous devions calculer la moyenne pour chaque colonne
print(data_math.shape)
print(data_math.sum(numeric_only=True)/data_math.shape[0])
print(data_math.sum(axis=1,numeric_only=True)/data_math.shape[1])

(303, 15)
age           5.986817
sex           0.683168
cp            0.966997
trtbps      131.623762
chol        246.264026
fbs           0.148515
restecg       0.528053
thalachh    149.646865
exng          0.326733
oldpeak       1.039604
slp           1.399340
caa           0.729373
thall         2.313531
output        0.544554
dtype: float64
0      36.218929
1      38.897400
2      34.616429
3      36.584921
4      43.389222
         ...    
298    34.412597
299    34.814410
300    33.086110
301    26.057762
302    36.731757
Length: 303, dtype: float64


In [67]:
# Plus simple comme ceci:
print(df[['age','chol']].quantile(0.25))
df[['age','chol']].loc[1:20].mean()

age      47.5
chol    211.0
Name: 0.25, dtype: float64


age      53.75
chol    246.95
dtype: float64

In [68]:
# Lien de corrélation
df.corr(numeric_only=True)

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output
age,1.000000,-0.098447,-0.068653,0.279351,0.213678,0.121308,-0.116211,-0.398522,0.096801,0.210013,-0.168814,0.276326,0.068001,-0.225439
sex,-0.098447,1.000000,-0.049353,-0.056769,-0.197912,0.045032,-0.058196,-0.044020,0.141664,0.096093,-0.030711,0.118261,0.210041,-0.280937
cp,-0.068653,-0.049353,1.000000,0.047608,-0.076904,0.094444,0.044421,0.295762,-0.394280,-0.149230,0.119717,-0.181053,-0.161736,0.433798
trtbps,0.279351,-0.056769,0.047608,1.000000,0.123174,0.177531,-0.114103,-0.046698,0.067616,0.193216,-0.121475,0.101389,0.062210,-0.144931
chol,0.213678,-0.197912,-0.076904,0.123174,1.000000,0.013294,-0.151040,-0.009940,0.067023,0.053952,-0.004038,0.070511,0.098803,-0.085239
fbs,0.121308,0.045032,0.094444,0.177531,0.013294,1.000000,-0.084189,-0.008567,0.025665,0.005747,-0.059894,0.137979,-0.032019,-0.028046
restecg,-0.116211,-0.058196,0.044421,-0.114103,-0.151040,-0.084189,1.000000,0.044123,-0.070733,-0.058770,0.093045,-0.072042,-0.011981,0.137230
thalachh,-0.398522,-0.044020,0.295762,-0.046698,-0.009940,-0.008567,0.044123,1.000000,-0.378812,-0.344187,0.386784,-0.213177,-0.096439,0.421741
exng,0.096801,0.141664,-0.394280,0.067616,0.067023,0.025665,-0.070733,-0.378812,1.000000,0.288223,-0.257748,0.115739,0.206754,-0.436757
oldpeak,0.210013,0.096093,-0.149230,0.193216,0.053952,0.005747,-0.058770,-0.344187,0.288223,1.000000,-0.577537,0.222682,0.210244,-0.430696


Nous avons vu comment créer des catégories. Comment calculer la moyenne par exemple en fonction de ces catégories?

Avec la fonction **groupby** comme en SQL.


In [69]:
df.groupby(by=['age_cat']).mean()

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output
age_cat,,,,,,,,,,,,,,
adult,52.625926,0.692593,0.955556,130.911111,244.959259,0.144444,0.533333,151.337037,0.325926,1.026296,1.414815,0.674074,2.318519,0.548148
senior,68.606061,0.606061,1.060606,137.454545,256.939394,0.181818,0.484848,135.818182,0.333333,1.148485,1.272727,1.181818,2.272727,0.515152


In [70]:
df.groupby(by=['age_cat','output'],as_index=False).mean()

,age_cat,output,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall
0,adult,0,55.073770,0.811475,0.467213,133.573770,251.286885,0.147541,0.450820,140.065574,0.540984,1.604918,1.180328,1.098361,2.549180
1,adult,1,50.608108,0.594595,1.358108,128.716216,239.743243,0.141892,0.601351,160.628378,0.148649,0.549324,1.608108,0.324324,2.128378
2,senior,0,68.250000,0.937500,0.562500,140.687500,249.562500,0.250000,0.437500,131.750000,0.625000,1.437500,1.062500,1.687500,2.500000
3,senior,1,68.941176,0.294118,1.529412,134.411765,263.882353,0.117647,0.529412,139.647059,0.058824,0.876471,1.470588,0.705882,2.058824


In [71]:
df.groupby(by=['age_cat','output']).mean().reset_index()

,age_cat,output,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall
0,adult,0,55.073770,0.811475,0.467213,133.573770,251.286885,0.147541,0.450820,140.065574,0.540984,1.604918,1.180328,1.098361,2.549180
1,adult,1,50.608108,0.594595,1.358108,128.716216,239.743243,0.141892,0.601351,160.628378,0.148649,0.549324,1.608108,0.324324,2.128378
2,senior,0,68.250000,0.937500,0.562500,140.687500,249.562500,0.250000,0.437500,131.750000,0.625000,1.437500,1.062500,1.687500,2.500000
3,senior,1,68.941176,0.294118,1.529412,134.411765,263.882353,0.117647,0.529412,139.647059,0.058824,0.876471,1.470588,0.705882,2.058824


In [96]:
df.drop(columns=['age_cat'],inplace=True)

In [97]:
# Il suffit de spécifier la colonne
df.loc[df['age'] <= 18, 'age_cat'] = 'child'
df.loc[(df['age'] > 18) & (df['age'] < 66), 'age_cat'] = 'adult'
df.loc[df['age'] > 65, 'age_cat'] = 'senior'
grouped_data = df.groupby(['age_cat','output']).mean(numeric_only=True)
display(grouped_data)
grouped_data.reset_index()

age       sex        cp      trtbps        chol  \
age_cat output                                                          
adult   0       61.942623  0.811475  0.467213  133.573770  251.286885   
        1       62.621622  0.594595  1.358108  128.716216  239.743243   
senior  0       68.250000  0.937500  0.562500  140.687500  249.562500   
        1       68.941176  0.294118  1.529412  134.411765  263.882353   

                     fbs   restecg    thalachh      exng   oldpeak       slp  \
age_cat output                                                                 
adult   0       0.147541  0.450820  140.065574  0.540984  1.604918  1.180328   
        1       0.141892  0.601351  160.628378  0.148649  0.549324  1.608108   
senior  0       0.250000  0.437500  131.750000  0.625000  1.437500  1.062500   
        1       0.117647  0.529412  139.647059  0.058824  0.876471  1.470588   

                     caa     thall  
age_cat output                      
adult   0       1.098361  2.549180  
        1       0.324324  2.128378  
senior  0       1.687500  2.500000  
        1       0.705882  2.058824

,age_cat,output,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall
0,adult,0,61.942623,0.811475,0.467213,133.573770,251.286885,0.147541,0.450820,140.065574,0.540984,1.604918,1.180328,1.098361,2.549180
1,adult,1,62.621622,0.594595,1.358108,128.716216,239.743243,0.141892,0.601351,160.628378,0.148649,0.549324,1.608108,0.324324,2.128378
2,senior,0,68.250000,0.937500,0.562500,140.687500,249.562500,0.250000,0.437500,131.750000,0.625000,1.437500,1.062500,1.687500,2.500000
3,senior,1,68.941176,0.294118,1.529412,134.411765,263.882353,0.117647,0.529412,139.647059,0.058824,0.876471,1.470588,0.705882,2.058824


In [98]:
grouped_data.index

MultiIndex([( 'adult', 0),
            ( 'adult', 1),
            ('senior', 0),
            ('senior', 1)],
           names=['age_cat', 'output'])

In [99]:
df.groupby(['age_cat','output']).agg({'age':['mean','max','min'],
                                        'chol':'max'})

age             chol
                     mean   max   min  max
age_cat output                            
adult   0       61.942623  65.0  58.0  409
        1       62.621622  65.0  58.0  417
senior  0       68.250000  77.0  66.0  322
        1       68.941176  76.0  66.0  564

In [100]:
# Nous n'avons pas la catégorie child, pourquoi?
# Value counts permet d'inventorier nos valeurs catégorielles
df['age_cat'].value_counts()

age_cat
adult     270
senior     33
Name: count, dtype: int64

In [101]:
df.drop_duplicates()

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output,age_cat
0,63.0,1,3,145,233,1,0,150,0,2.3,0,0,1,1,adult
1,63.0,1,2,130,250,0,1,187,0,3.5,0,0,2,1,adult
2,63.0,0,1,130,204,0,0,172,0,1.4,2,0,2,1,adult
3,63.0,1,1,120,236,0,1,178,0,0.8,2,0,2,1,adult
4,63.0,0,0,120,354,0,1,163,1,0.6,2,0,2,1,adult
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
298,63.0,0,0,140,241,0,1,123,1,0.2,1,0,3,0,adult
299,63.0,1,3,110,264,0,1,132,0,1.2,1,0,3,0,adult
300,68.0,1,0,144,193,1,1,141,0,3.4,1,2,3,0,senior
301,63.0,1,0,130,131,0,1,115,1,1.2,1,1,3,0,adult


In [102]:
display(df.duplicated())
df[df.duplicated()]

0      False
1      False
2      False
3      False
4      False
       ...  
298    False
299    False
300    False
301    False
302    False
Length: 303, dtype: bool

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output,age_cat
164,63.0,1,2,138,175,0,1,173,0,0.0,2,4,2,1,adult


In [103]:
# Vérifions
print(df['age'].min())
df['age'].max()


58.0


np.float64(77.0)

In [106]:
# La fonction describe permet d'obtenir une vision très macro de vos données
display(df.describe(include='all').T)
df.describe(include='all')

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
age,303.0,NaN,NaN,NaN,63.0,2.6818,58.0,63.0,63.0,63.0,77.0
sex,303.0,NaN,NaN,NaN,0.683168,0.466011,0.0,0.0,1.0,1.0,1.0
cp,303.0,NaN,NaN,NaN,0.966997,1.032052,0.0,0.0,1.0,2.0,3.0
trtbps,303.0,NaN,NaN,NaN,131.623762,17.538143,94.0,120.0,130.0,140.0,200.0
chol,303.0,NaN,NaN,NaN,246.264026,51.830751,126.0,211.0,240.0,274.5,564.0
fbs,303.0,NaN,NaN,NaN,0.148515,0.356198,0.0,0.0,0.0,0.0,1.0
restecg,303.0,NaN,NaN,NaN,0.528053,0.52586,0.0,0.0,1.0,1.0,2.0
thalachh,303.0,NaN,NaN,NaN,149.646865,22.905161,71.0,133.5,153.0,166.0,202.0
exng,303.0,NaN,NaN,NaN,0.326733,0.469794,0.0,0.0,0.0,1.0,1.0
oldpeak,303.0,NaN,NaN,NaN,1.039604,1.161075,0.0,0.0,0.8,1.6,6.2


,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output,age_cat
count,303.0000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303.000000,303
unique,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2
top,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,adult
freq,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,270
mean,63.0000,0.683168,0.966997,131.623762,246.264026,0.148515,0.528053,149.646865,0.326733,1.039604,1.399340,0.729373,2.313531,0.544554,NaN
std,2.6818,0.466011,1.032052,17.538143,51.830751,0.356198,0.525860,22.905161,0.469794,1.161075,0.616226,1.022606,0.612277,0.498835,NaN
min,58.0000,0.000000,0.000000,94.000000,126.000000,0.000000,0.000000,71.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,NaN
25%,63.0000,0.000000,0.000000,120.000000,211.000000,0.000000,0.000000,133.500000,0.000000,0.000000,1.000000,0.000000,2.000000,0.000000,NaN
50%,63.0000,1.000000,1.000000,130.000000,240.000000,0.000000,1.000000,153.000000,0.000000,0.800000,1.000000,0.000000,2.000000,1.000000,NaN
75%,63.0000,1.000000,2.000000,140.000000,274.500000,0.000000,1.000000,166.000000,1.000000,1.600000,2.000000,1.000000,3.000000,1.000000,NaN


## <u>5. Comment gérer les valeurs manquantes:</u>

In [81]:
# D'abord modifions notre dataset de sorte à avoir des données manquantes
# nan correspond à Null, aux valeurs manquantes
# Ici l'âge pour toutes les personnes de moins de 58 ans sera à NaN
df.loc[df['age'] < 58, 'age'] = np.nan
df

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output,age_cat
0,63.0,1,3,145,233,1,0,150,0,2.3,0,0,1,1,adult
1,NaN,1,2,130,250,0,1,187,0,3.5,0,0,2,1,adult
2,NaN,0,1,130,204,0,0,172,0,1.4,2,0,2,1,adult
3,NaN,1,1,120,236,0,1,178,0,0.8,2,0,2,1,adult
4,NaN,0,0,120,354,0,1,163,1,0.6,2,0,2,1,adult
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
298,NaN,0,0,140,241,0,1,123,1,0.2,1,0,3,0,adult
299,NaN,1,3,110,264,0,1,132,0,1.2,1,0,3,0,adult
300,68.0,1,0,144,193,1,1,141,0,3.4,1,2,3,0,senior
301,NaN,1,0,130,131,0,1,115,1,1.2,1,1,3,0,adult


In [108]:
# D'abord il faut les identifier
# isna() permet de renvoyer True ou False si les données sont manquantes ou non
# Ensuite sum() additionne le tout. False étant == 0 et True == 1 on peut en déduire le nombre d'élément manquant

df.isna().sum()

age         0
sex         0
cp          0
trtbps      0
chol        0
fbs         0
restecg     0
thalachh    0
exng        0
oldpeak     0
slp         0
caa         0
thall       0
output      0
age_cat     0
dtype: int64

In [83]:
# fillna() permet de remplacer des valeurs manquantes par une valeur choisie
# ici on choisir de remplacer par la moyenne
df['age'] = df['age'].fillna(df['age'].mean())
df

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output,age_cat
0,63.0,1,3,145,233,1,0,150,0,2.3,0,0,1,1,adult
1,63.0,1,2,130,250,0,1,187,0,3.5,0,0,2,1,adult
2,63.0,0,1,130,204,0,0,172,0,1.4,2,0,2,1,adult
3,63.0,1,1,120,236,0,1,178,0,0.8,2,0,2,1,adult
4,63.0,0,0,120,354,0,1,163,1,0.6,2,0,2,1,adult
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
298,63.0,0,0,140,241,0,1,123,1,0.2,1,0,3,0,adult
299,63.0,1,3,110,264,0,1,132,0,1.2,1,0,3,0,adult
300,68.0,1,0,144,193,1,1,141,0,3.4,1,2,3,0,senior
301,63.0,1,0,130,131,0,1,115,1,1.2,1,1,3,0,adult


On pourrait également supprimer les valeurs manquantes.

Attention à ne pas trop supprimer de lignes ou de colonnes.

In [109]:
df[['age']].sort_values()

TypeError: DataFrame.sort_values() missing 1 required positional argument: 'by'

In [111]:
df[['age','cp','sex']].sort_values(by=['age','cp'],ascending=[False,True])

,age,cp,sex
238,77.0,0,1
144,76.0,2,0
129,74.0,1,0
151,71.0,0,0
25,71.0,1,0
...,...,...,...
64,58.0,2,1
79,58.0,2,1
173,58.0,2,1
183,58.0,2,1


In [ ]:
df.loc[df['age'] == 63, 'age'] = np.nan
df = df.dropna(subset=['age'])
df.sort_values(by=['age','chol'],ascending=[True,False])

,age,sex,cp,trtbps,chol,fbs,restecg,thalachh,exng,oldpeak,slp,caa,thall,output,age_cat
16,58.0,0,2,120,340,0,1,172,0,0.0,2,0,2,1,adult
278,58.0,0,1,136,319,1,0,152,0,0.0,2,2,2,0,adult
291,58.0,1,0,114,318,0,2,140,0,4.4,0,3,1,0,adult
236,58.0,1,0,125,300,0,0,171,0,0.0,2,2,3,0,adult
172,58.0,1,1,120,284,0,0,160,0,1.8,1,0,2,0,adult
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
60,71.0,0,2,110,265,1,0,130,0,0.0,2,1,2,1,senior
151,71.0,0,0,112,149,0,1,125,0,1.6,1,0,2,1,senior
129,74.0,0,1,120,269,0,0,121,1,0.2,2,1,2,1,senior
144,76.0,0,2,140,197,0,2,116,0,1.1,1,0,2,1,senior


Nous sommes passés de 303 lignes à 114, ce qui fait beaucoup.

Il serait plus sage de supprimer la colonne age plutôt que de supprimer chaque ligne où il y a une valeur manquante.

## <u>6. Comment join plusieurs datasets </U>

Fonctionnement identique au join SQL

Illustation avec le pd.read_clipboard()

PIB: https://fr.wikipedia.org/wiki/Produit_int%C3%A9rieur_brut

Population: https://fr.wikipedia.org/wiki/Liste_des_pays_par_population

In [122]:
pib = pd.read_clipboard()
pib

,Pays,2015,2020,2025
0,Drapeau du Luxembourg Luxembourg,279,255,239
1,Drapeau des États-Unis États-Unis,147,135,138
2,Drapeau des Pays-Bas Pays-Bas,132,131,133
3,Drapeau de l'Allemagne Allemagne,126,124,115
4,Drapeau de la Belgique Belgique,119,118,114
5,Drapeau de l’Union européenne Union européenne...,100,100,100
6,Drapeau du Royaume-Uni Royaume-Uni,111,102,99
7,Drapeau de la France France,106,103,98
8,Drapeau de l'Italie Italie,97,93,96
9,Drapeau de l'Espagne Espagne,91,82,92


In [123]:

pib['Pays'] = pib['Pays'].apply(lambda x: x.split()[-1])
pib

,Pays,2015,2020,2025
0,Luxembourg,279,255,239
1,États-Unis,147,135,138
2,Pays-Bas,132,131,133
3,Allemagne,126,124,115
4,Belgique,119,118,114
5,pays),100,100,100
6,Royaume-Uni,111,102,99
7,France,106,103,98
8,Italie,97,93,96
9,Espagne,91,82,92


In [124]:
pib = pib.drop(5)
pib

,Pays,2015,2020,2025
0,Luxembourg,279,255,239
1,États-Unis,147,135,138
2,Pays-Bas,132,131,133
3,Allemagne,126,124,115
4,Belgique,119,118,114
6,Royaume-Uni,111,102,99
7,France,106,103,98
8,Italie,97,93,96
9,Espagne,91,82,92
10,Japon,108,91,85


In [125]:
pib['Pays'].unique()

<StringArray>
[ 'Luxembourg',  'États-Unis',    'Pays-Bas',   'Allemagne',    'Belgique',
 'Royaume-Uni',      'France',      'Italie',     'Espagne',       'Japon',
     'Pologne',     'Turquie']
Length: 12, dtype: str

In [128]:
pib

,Pays,2015,2020,2025
0,Luxembourg,279,255,239
1,États-Unis,147,135,138
2,Pays-Bas,132,131,133
3,Allemagne,126,124,115
4,Belgique,119,118,114
6,Royaume-Uni,111,102,99
7,France,106,103,98
8,Italie,97,93,96
9,Espagne,91,82,92
10,Japon,108,91,85


In [ ]:
# Depivotage
pib.melt(id_vars=['Pays'],value_vars=pib.drop(columns=['Pays']).columns,var_name='Annee').sort_values(by=['Pays'])

,Pays,Annee,value
14,Allemagne,2019,122
3,Allemagne,2014,128
25,Allemagne,2024,116
2,Belgique,2014,121
13,Belgique,2019,117
24,Belgique,2024,117
18,Espagne,2019,91
29,Espagne,2024,91
7,Espagne,2014,90
27,France,2024,98


In [129]:
pib.pivot_table() # Inverse d'un melt

ValueError: No group keys passed!

In [130]:
url = 'https://fr.wikipedia.org/wiki/Liste_des_pays_par_population'
response = rq.get(url,headers={'User-Agent': 'Mozilla/5.0'})
data = pd.read_html(StringIO(response.text))
population = data[0]
population

,Rang (2023),Pays ou territoire,Population au 1er juillet 2023,Population projetée au 1er juillet 2026
0,-,Monde,8 091 734 930,8 300 678 395
1,1,Inde[a],1 438 069 596,1 476 625 576
2,2,Chine[b],1 422 584 933,1 412 914 089
3,3,États-Unis[c],343 477 335,349 035 494
4,4,Indonésie,281 190 067,287 886 782
...,...,...,...,...
233,-,Montserrat,4 420,4 328
234,-,Îles Malouines,3 477,3 465
235,-,Tokelau,2 397,2 691
236,199,Niue,1 817,1 822


In [131]:
population['Pays ou territoire'] = population['Pays ou territoire'].str.replace(r'\[.*\]', '',regex=True).str.strip()

In [132]:
population

,Rang (2023),Pays ou territoire,Population au 1er juillet 2023,Population projetée au 1er juillet 2026
0,-,Monde,8 091 734 930,8 300 678 395
1,1,Inde,1 438 069 596,1 476 625 576
2,2,Chine,1 422 584 933,1 412 914 089
3,3,États-Unis,343 477 335,349 035 494
4,4,Indonésie,281 190 067,287 886 782
...,...,...,...,...
233,-,Montserrat,4 420,4 328
234,-,Îles Malouines,3 477,3 465
235,-,Tokelau,2 397,2 691
236,199,Niue,1 817,1 822


In [135]:
print(list(population))
print(list(pib))
finaldf = pib.merge(population, how='left', left_on='Pays', right_on='Pays ou territoire')
finaldf


['Rang (2023)', 'Pays ou territoire', 'Population au 1er juillet 2023', 'Population projetée au 1er juillet 2026']
['Pays', '2015', '2020', '2025']


,Pays,2015,2020,2025,Rang (2023),Pays ou territoire,Population au 1er juillet 2023,Population projetée au 1er juillet 2026
0,Luxembourg,279,255,239,165,Luxembourg,665 098,687 448
1,États-Unis,147,135,138,3,États-Unis,343 477 335,349 035 494
2,Pays-Bas,132,131,133,70,Pays-Bas,18 092 524,18 448 775
3,Allemagne,126,124,115,19,Allemagne,84 548 231,83 644 258
4,Belgique,119,118,114,81,Belgique,11 712 893,11 774 642
5,Royaume-Uni,111,102,99,21,Royaume-Uni,68 682 962,69 931 528
6,France,106,103,98,23,France,66 438 822,66 746 401
7,Italie,97,93,96,25,Italie,59 499 453,58 926 166
8,Espagne,91,82,92,32,Espagne,47 911 579,47 850 793
9,Japon,108,91,85,12,Japon,124 370 947,122 427 731


In [136]:
finaldf = finaldf.rename(columns={'Rang (2023)':'Rang population'})
finaldf

,Pays,2015,2020,2025,Rang population,Pays ou territoire,Population au 1er juillet 2023,Population projetée au 1er juillet 2026
0,Luxembourg,279,255,239,165,Luxembourg,665 098,687 448
1,États-Unis,147,135,138,3,États-Unis,343 477 335,349 035 494
2,Pays-Bas,132,131,133,70,Pays-Bas,18 092 524,18 448 775
3,Allemagne,126,124,115,19,Allemagne,84 548 231,83 644 258
4,Belgique,119,118,114,81,Belgique,11 712 893,11 774 642
5,Royaume-Uni,111,102,99,21,Royaume-Uni,68 682 962,69 931 528
6,France,106,103,98,23,France,66 438 822,66 746 401
7,Italie,97,93,96,25,Italie,59 499 453,58 926 166
8,Espagne,91,82,92,32,Espagne,47 911 579,47 850 793
9,Japon,108,91,85,12,Japon,124 370 947,122 427 731


In [140]:
finaldf.to_csv("country_pib.csv")

## <u>7. connection db </U>

<br>

Créer une connecteur sur base d'une URL avec comme information:
- le SGBD
- username+password
- adresse réseau
- le nom de la db à laquelle se connecter

E.g.:
- postgresql://username:password@localhost:5432/db_name # localhost=127.0.0.1
- mssql+pyodbc://username:password@localhost:1433/db_name
- mssql+pyodbc://server_name/db_name?trusted_connection=yes # auth windows
- mssql+pyodbc://server_name/db_name?trusted_connection=yes&driver=ODBC+Driver+17+for+SQL+Server # Avec driver ODBC spécifique

In [143]:
from dotenv import load_dotenv

In [141]:
sgbd = 'postgresql'
server_name = r'ERNEST\DATA'
db_name = 'DBPandas'
driver = 'ODBC+Driver+17+for+SQL+Server'

url = f'{sgbd}://{server_name}/{db_name}?trusted_connection=yes&driver={driver}'
con = create_engine(url)

sql_query = "SELECT * FROM student WHERE year_result > 15"
data = pd.read_sql(sql_query,con)

con.dispose() # release con

data

OperationalError: (pyodbc.OperationalError) ('08001', '[08001] [Microsoft][ODBC Driver 17 for SQL Server]SQL Server Network Interfaces: Error Locating Server/Instance Specified [xFFFFFFFF].  (-1) (SQLDriverConnect); [08001] [Microsoft][ODBC Driver 17 for SQL Server]Login timeout expired (0); [08001] [Microsoft][ODBC Driver 17 for SQL Server]A network-related or instance-specific error has occurred while establishing a connection to SQL Server. Server is not found or not accessible. Check if instance name is correct and if SQL Server is configured to allow remote connections. For more information see SQL Server Books Online. (-1)')
(Background on this error at: https://sqlalche.me/e/20/e3q8)